### PR TITLE
Complex multiplication and conjugate contrib ops 

### PR DESCRIPTION
**Pull request by duli2012**
**Description**: Describe your changes.
Implemented two contrib ops: complex multiplication and complex conjugate. Both are standard mathematic computations: 
https://mathworld.wolfram.com/ComplexMultiplication.html
https://mathworld.wolfram.com/ComplexConjugate.html

We decide to implement them as contrib ops rather than custom ops because these are standard mathematic computation, which is likely to be used by other models.

**Motivation and Context**
- Why is this change required? What problem does it solve?
A vision model need them

### CODE REVIEW

**Comment by duli2012** [comment]
Please don't review this PR until PR 3381 got merged.

**Comment by pranavsharma** [comment]
Is there a rationale for adding these as contrib ops? Can the vision team not use and register custom ops?

**Comment by snnn** [comment]
> Is there a rationale for adding these as contrib ops? Can the vision team not use and register custom ops?

When I added the cdist op, @faxu asked me to send an email to our internal onnxoperatortriage team and to get an approval before checking the contrib op to ORT master. I suggest we should continue to follow such process.

**Comment by duli2012** [comment]
> Can you add the rationale for adding this as a contrib op and not a custom op in the PR description?

Done.

File: onnxruntime/contrib_ops/cuda/math/complex_mul.cc [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+      kMSDomain,                                                  \
+      1,                                                          \
+      T,                                                          \
+      kCudaExecutionProvider,                                     \
+      KernelDefBuilder()                                          \
+          .TypeConstraint("T", DataTypeImpl::GetTensorType<T>()), \
+      ComplexMulConj<T>);
+
+REGISTER_KERNEL_TYPED(float)
+REGISTER_KERNEL_TYPED(double)
```
Comment (HectorSVC) [inline onnxruntime/contrib_ops/cuda/math/complex_mul.cc:31]: >double [](start = 22, length = 6)

Is double required? didn't see it in BuildKernelCreateInfo

File: onnxruntime/test/contrib_ops/element_wise_ops_test.cc [hunk 2]
```diff
@@ -114,5 +114,52 @@ TEST(BiasGeluTest, Two_One_Dim) {
   RunBiasGeluTest(input_a_data, input_b_data, {2, 4}, {4});
 }
 
+TEST(MathOpTest, ComplexMul) {
```
Comment (HectorSVC) [inline onnxruntime/test/contrib_ops/element_wise_ops_test.cc:117]: >TEST [](start = 0, length = 4)

test case to cover fp16 also?

File: onnxruntime/contrib_ops/cuda/math/complex_mul.cc [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+REGISTER_KERNEL_TYPED(MLFloat16)
+
+template <typename T>
+Status ComplexMul<T>::ComputeInternal(OpKernelContext* context) const {
+  for (int index = 0; index < context->InputCount(); ++index) {
+    const Tensor* input = context->Input<Tensor>(index);
+    TensorShape shape = input->Shape();
+    int64_t last_dimension = shape[shape.NumDimensions() - 1];
+    ORT_ENFORCE(last_dimension == 2, "The input_", index, " last demension is supposed to be 2, but get ", last_dimension);
```
Comment (HectorSVC) [inline onnxruntime/contrib_ops/cuda/math/complex_mul.cc:40]: >demension [](start = 65, length = 9)

dimension

File: onnxruntime/contrib_ops/cuda/math/complex_mul.cc [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  return Status::OK();
+}
+
+template <typename T>
+Status ComplexMulConj<T>::ComputeInternal(OpKernelContext* context) const {
+  for (int index = 0; index < context->InputCount(); ++index) {
+    const Tensor* input = context->Input<Tensor>(index);
+    TensorShape shape = input->Shape();
+    int64_t last_dimension = shape[shape.NumDimensions() - 1];
+    ORT_ENFORCE(last_dimension == 2, "The input_", index, " last demension is supposed to be 2, but get ", last_dimension);
```
Comment (HectorSVC) [inline onnxruntime/contrib_ops/cuda/math/complex_mul.cc:68]: >demension [](start = 65, length = 9)

dimension

File: onnxruntime/contrib_ops/cuda/math/complex_mul.cc [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+      &prepare.rhs_padded_strides,
+      reinterpret_cast<const typename ToCudaType<T>::MappedType*>(prepare.rhs_tensor->template Data<T>()),
+      &prepare.fdm_output_strides,
+      prepare.fdm_H,
+      prepare.fdm_C,
+      reinterpret_cast<typename ToCudaType<T>::MappedType*>(prepare.output_tensor->template MutableData<T>()),
+      prepare.output_tensor->Shape().Size(),
+      prepare.lhs_tensor->Shape().Size(),
+      prepare.rhs_tensor->Shape().Size(),
+      false);
```
Comment (HectorSVC) [inline onnxruntime/contrib_ops/cuda/math/complex_mul.cc:58]: >false [](start = 6, length = 5)

if this bool flag is the only difference between ComplexMul and ComplexMulConj, you can merge them into one class, put  the flag into template< bool is_conj> SuperClass...

File: onnxruntime/contrib_ops/cuda/math/complex_mul.cc (hunk 3 above)
Comment (duli2012) [reply onnxruntime/contrib_ops/cuda/math/complex_mul.cc]: fixed

File: onnxruntime/contrib_ops/cuda/math/complex_mul.cc (hunk 4 above)
Comment (duli2012) [reply onnxruntime/contrib_ops/cuda/math/complex_mul.cc]: fixed

File: onnxruntime/contrib_ops/cuda/math/complex_mul.cc (hunk 5 above)
Comment (duli2012) [reply onnxruntime/contrib_ops/cuda/math/complex_mul.cc]: done

File: onnxruntime/test/contrib_ops/element_wise_ops_test.cc (hunk 2 above)
Comment (duli2012) [reply onnxruntime/test/contrib_ops/element_wise_ops_test.cc:117]: added

File: onnxruntime/contrib_ops/cuda/math/complex_mul.cc (hunk 1 above)
Comment (duli2012) [reply onnxruntime/contrib_ops/cuda/math/complex_mul.cc]: removed

**Pull request merged**